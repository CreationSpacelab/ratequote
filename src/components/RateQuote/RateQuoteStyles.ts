import styled from "styled-components";

import { fadeIn } from "../core/animation";
import { theme } from "../core/colors";

export const Column = styled.div`
  flex: 1;
  margin: 16px auto;
  max-width: 16.6667%;
  padding: 0 16px;
`;

export const LenderName = styled(Column)`
  overflow-x: scroll;
  white-space: nowrap;
`;

export const RateQuoteRow = styled.div`
  border-bottom: 1px solid ${theme.lightgray};
  display: flex;
  align-items: center;
  letter-spacing: 0.5px;
  margin: 0 auto;
  padding: 16px 8px;
  position: relative;
  animation: ${fadeIn} 450ms ease-in forwards;
`;

export const TableHeaderRow = styled(RateQuoteRow)`
  padding: 8px;
  animation: none;

  ${Column} {
    color: ${theme.gray};
    font-size: 14px;
    letter-spacing: 1.2px;
    text-transform: uppercase;
  }
`;

export const ToolTip = styled.div`
  display: inline-block;
  transition: all 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);

  &:hover {
    > div {
      visibility: visible;
    }
  }
  div {
    background-color: ${theme.blue};
    border-radius: 3px;
    display: block;
    color: ${theme.white};
    font-size: 13px;
    padding: 12px;
    position: absolute;
    left: 20px;
    bottom: -10px;
    text-align: center;
    visibility: hidden;
    width: fit-content;
    z-index: 1000;
  }
`;
