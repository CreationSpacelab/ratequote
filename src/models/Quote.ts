import currency from "currency.js";

import { QuoteType } from "../quoteTypes";
/**
 *  @class Quote
 *  @property {Function} interestRateFormatted
 *  @property {Function} monthlyPaymentFormatted
 *  @property {Function} aprFormatted
 */
class Quote {
  /** Member variables */
  lenderName: string;
  loanType: string;
  interestRate: number;
  closingCosts: number;
  monthlyPayment: number;
  apr: number;

  /**
   * Construct a Quote of type QuoteType.
   * @param {Object} payload
   * @typedef {Object} QuoteType
   */
  constructor(payload: QuoteType) {
    this.lenderName = payload.lenderName;
    this.loanType = payload.loanType;
    this.interestRate = payload.interestRate;
    this.closingCosts = payload.closingCosts;
    this.monthlyPayment = payload.monthlyPayment;
    this.apr = payload.apr;
  }

  /**
   * Format the 'interestRate' data to 2 decimal points.
   * @return {string} interestRate formatted to String.
   */
  interestRateFormatted(): string {
    return `${this.interestRate.toFixed(3)}%`;
  }

  /**
   * Format the 'monthlyPayment' data to 2 decimal points.
   * @return {string} monthlyPayment formatted to String.
   */
  closingCostsFormatted(): string {
    return currency(Math.round(this.closingCosts), {
      fromCents: false,
      precision: 0,
    }).format();
  }
  /**
   * Format the 'monthlyPayment' data to 2 decimal points.
   * @return {string} monthlyPayment formatted to String.
   */
  monthlyPaymentFormatted(): string {
    return currency(Math.round(this.monthlyPayment), {
      fromCents: false,
      precision: 0,
    }).format();
  }

  /**
   * Format the 'apr' data to 2 decimal points.
   * @return {string} apr formatted to String.
   */
  aprFormatted(): string {
    return `${this.apr.toFixed(3)}%`;
  }
}

export default Quote;
