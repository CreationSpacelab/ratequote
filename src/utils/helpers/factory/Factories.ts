import { Factory } from "rosie";

Factory.define("RateQuote").attrs({
  lenderName: "TFB Federal Credit Union",
  loanType: "7/1 ARM",
  interestRate: 3.375,
  closingCosts: 5500,
  monthlyPayment: 1989.4329462037347,
  apr: 3.472295213050529,
  interestRateFormatted: "3.375%",
  monthlyPaymentFormatted: "$1,989",
  aprFormatted: "3.472%",
});

export default Factory;
