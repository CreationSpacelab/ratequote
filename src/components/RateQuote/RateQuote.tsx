import React from "react";

import Quote from "../../models/Quote";
import { QuoteType } from "../../quoteTypes";

import { Column, RateQuoteRow, LenderName, ToolTip } from "./RateQuoteStyles";

const LENDER_CHAR_LIMIT = 13

interface RateQuoteProps {
  quote: QuoteType
}

const RateQuote: React.FC<RateQuoteProps> = ({ quote }: RateQuoteProps) => {
  const rateQuote = new Quote(quote);

  // Limit the amount of lenderName characters displayed
  const lenderName: string = rateQuote.lenderName.length > LENDER_CHAR_LIMIT
    ? `${rateQuote.lenderName.substring(0, LENDER_CHAR_LIMIT)}...`
    : rateQuote.lenderName

  return (
    <>
      <RateQuoteRow>
        <LenderName data-testid="lenderName">
          <ToolTip>
            {lenderName}
            <div>{rateQuote.lenderName}</div>
          </ToolTip>
        </LenderName>
        <Column data-testid="loanType">{rateQuote.loanType}</Column>
        <Column data-testid="interestRateFormatted">{rateQuote.interestRateFormatted()}</Column>
        <Column data-testid="closingCosts">{rateQuote.closingCostsFormatted()}</Column>
        <Column data-testid="monthlyPaymentFormatted">{rateQuote.monthlyPaymentFormatted()}</Column>
        <Column data-testid="aprFormatted">{rateQuote.aprFormatted()}</Column>
      </RateQuoteRow>
    </>
  );
};

export default React.memo(RateQuote)

