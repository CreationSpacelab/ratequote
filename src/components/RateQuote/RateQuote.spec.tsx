import React from "react";

import { render } from "@testing-library/react";

import Quote from "../../models/Quote";
import Factories from "../../utils/helpers/factory/Factories"

import RateQuote from "./RateQuote";

const baseQuote: Quote = Factories.build("RateQuote")

describe("RateQuote", () =>{

  test("renders app correctly", () => {
    const { container } = render(<RateQuote quote={baseQuote} />);
    expect(container).toMatchSnapshot()
  });

  test("displays the interest rate formatted", () => {
    const { getByText } = render(<RateQuote quote={baseQuote} />);
    getByText("3.375%", { exact: false })
  });

  test("displays the monthly payment formatted", () => {
    const { getByText } = render(<RateQuote quote={baseQuote} />);
    getByText("$1,989", { exact: false })
  });

  test("displays the apr formatted", () => {
    const { getByText } = render(<RateQuote quote={baseQuote} />);
    getByText("3.472%", {exact: false})
  });
})
