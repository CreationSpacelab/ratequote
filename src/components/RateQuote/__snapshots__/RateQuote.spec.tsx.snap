// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateQuote renders app correctly 1`] = `
<div>
  <div
    class="sc-dlfnbm huaKEx"
  >
    <div
      class="sc-bdfBwQ sc-gsTCUz gIxYza bKuXft"
      data-testid="lenderName"
    >
      <div
        class="sc-eCssSg ffHURk"
      >
        TFB Federal C...
        <div>
          TFB Federal Credit Union
        </div>
      </div>
    </div>
    <div
      class="sc-bdfBwQ gIxYza"
      data-testid="loanType"
    >
      7/1 ARM
    </div>
    <div
      class="sc-bdfBwQ gIxYza"
      data-testid="interestRateFormatted"
    >
      3.375%
    </div>
    <div
      class="sc-bdfBwQ gIxYza"
      data-testid="closingCosts"
    >
      $5,500
    </div>
    <div
      class="sc-bdfBwQ gIxYza"
      data-testid="monthlyPaymentFormatted"
    >
      $1,989
    </div>
    <div
      class="sc-bdfBwQ gIxYza"
      data-testid="aprFormatted"
    >
      3.472%
    </div>
  </div>
</div>
`;
